### PR TITLE
Change Set#delete to return Bool

### LINKED ISSUE #9591: Modify Set#delete (and possibily Set#add)
`Set(T)#delete` returns `self`, it will be more useful to return either `T?` or `Bool` to know if the object was present in the set.

We may also consider modifying `Set(T)#add` and `Set(T)#add?`, as suggested by @asterite .

Note that `Set(T)#<<` calls `Set(T).add` (similar than `Array(T)#<<` calls `Array(T)#push`), returns self, which is useful for chaining additions.

Related PR: https://github.com/crystal-lang/crystal/pull/9590

### PR DESCRIPTION
**Pull request by j8r**
It adds a simple and efficient way to know if the value was present or not.

### Design explanations

Having `Set#delete` returns `object` or `nil` is consistent with other `#delete` methods in the stdlib:
- `Hash#delete(object)` returns the `value` if `object` was present.
- `Array#delete_at(index)` returns the removed `value` at the given index, or raises an exception.

However, `Deque#delete(object)` returns a boolean (maybe it would be better to return a count of deleted objects, but not the topic).

Other options are:
- add `Set#delete?(object) : Bool` to have parity with `Set#add?`, but it can be considered redundant
- have `Set#delete(object)` return a `Bool`. Why not, but could be confusing: `#add` does not return a `Bool` (but `#add?` does)

Side note: I don't think `Set#add` and `Set#add?` are redundant, because `Set#<<` is using `Set#add`, and one can easily chain object additions like `set << one << second << third`.

Closes #9591

### CODE REVIEW

**Comment by asterite** [comment]
Unless `nil` is a valid value in the set ;-)

File: src/set.cr [hunk 1]
```diff
@@ -136,17 +136,17 @@ struct Set(T)
...
+  # Deletes the *object* and returns it, otherwise returns nil.
   #
   # ```
   # s = Set{1, 5}
   # s.includes? 5 # => true
-  # s.delete 5
+  # s.delete 5    # => 5
   # s.includes? 5 # => false
   # ```
   def delete(object)
```
Comment (asterite) [inline src/set.cr:147]:
```suggestion
  def delete(object) : T?
```

File: src/set.cr [hunk 2]
```diff
@@ -136,17 +136,17 @@ struct Set(T)
     @hash.has_key?(object)
   end
 
-  # Removes the *object* from the set and returns `self`.
+  # Deletes the *object* and returns it, otherwise returns nil.
```
Comment (Sija) [inline src/set.cr:139]:
```suggestion
  # Removes the *object* from the set - if found - and returns it, otherwise returns `nil`.
```

Comment (j8r) [reply src/set.cr:139]: I don't think is is necessary to add `if found`. It is logical to only remove an object if it exists, the opposite is not possible. `Hash#delete` doesn't have it.
Otherwise, valid suggestion.

File: src/set.cr (hunk 1 above)
Comment (j8r) [reply src/set.cr:147]: Ha, I didn't expect you suggesting to add return types :smile: .
I am not sure it is a good idea, `Hash` doesn't have this type restrictions.

File: src/set.cr (hunk 2 above)
Comment (asterite) [reply src/set.cr]: Hehe. I know this is safe. I knew it wasn't safe (because I implemented the compiler!) to add type restrictions in a block argument. So they are different things.

Comment (Sija) [reply src/set.cr]: You removed the part saying where you remove this object from.

File: src/set.cr (hunk 1 above)
Comment (j8r) [reply src/set.cr:147]: I guess it is safe, more than the type restriction of `object` at least.

File: src/set.cr [hunk 3]
```diff
@@ -311,7 +311,9 @@ struct Set(T)
   # ```
   def subtract(other : Enumerable)
     other.each do |value|
-      delete value
+      if value.is_a? T
+        delete value
+      end
```
Comment (asterite) [inline src/set.cr:316]: This seems to be unrelated to this PR, and also unnecessary.

Comment (j8r) [reply src/set.cr:316]: This is necessary, see why the previous builds failed.

Comment (j8r) [reply src/set.cr:316]: If `value` is not a `T`, `Set#delete(value)` won't return `T` and will fail to compile.

Comment (asterite) [reply src/set.cr:316]: Oh, so I think we should revert the type restriction, not add more code. You were right, it does break things.

Comment (asterite) [reply src/set.cr:316]: Actually, returning `object` is wrong. It should delete the object returned by `@hash.delete`.

File: src/set.cr [hunk 4]
```diff
@@ -136,17 +136,17 @@ struct Set(T)
...
-  # s.delete 5
+  # s.delete 5    # => 5
   # s.includes? 5 # => false
   # ```
-  def delete(object)
-    @hash.delete(object)
-    self
+  def delete(object) : T?
+    @hash.delete(object) { return }
+    object
```
Comment (asterite) [inline src/set.cr:149]:
```suggestion
    @hash.delete(object) { return nil }
```

File: src/set.cr (hunk 3 above)
Comment (j8r) [reply src/set.cr]: The suggestion does not really work. If the object is present, it will now return `nil` - Set is a hash of nils.

Comment (asterite) [reply src/set.cr]: Then I think this PR is wrong.

Comment (asterite) [reply src/set.cr]: Try this:

```crystal
s = Set{1, 2, 3}
s.delete(1.0) # => what does this return? It should be 1. If it returns 1.0 it's wrong.
```

Comment (j8r) [reply src/set.cr]: The issue will be exactly the same if we try to add a return type to `Hash#delete(key)`.

Comment (asterite) [reply src/set.cr]: No, because there we have the value.

Comment (asterite) [reply src/set.cr]: I think it makes sense to return `Bool`, not `T?`.

Comment (j8r) [reply src/set.cr]: In your example, `nil` will be returned. I am ok for `Bool`, no strong opinion on this.

**Comment by asterite** [comment]
I think we should make `delete` return `Bool`, and remove `add?` and make `add` return `Bool`.

But as usual, I think it's better to discuss these things in an issue, not in a PR.

File: src/set.cr (hunk 3 above)
Comment (j8r) [reply src/set.cr]: There is also another possible fix: add `return nil if !object.is_a? T`.

Comment (asterite) [reply src/set.cr]: In the case of `Array` and `Hash`, when you delete an index or a key you don't know what the underlying value is. Sometimes it's useful to know that.

In the case of `Set`, when you delete an object, that's the object that was already in the Set, so you don't learn anything new by returning that same object. It's more useful to know if the object was there or not, and for that you can use a `Bool`.

Comment (j8r) [reply src/set.cr]: One thing to consider is the object to ask to delete, passed to `#delete`, may not be exactly the same has the one stored.
They have the same hash of course. It may be not important after all, because it can be considered as a bad practice to have two different objects with the same hash.

File: src/set.cr [hunk 5]
```diff
@@ -136,17 +136,18 @@ struct Set(T)
...
+  # s.delete 5    # => 5
   # s.includes? 5 # => false
   # ```
-  def delete(object)
-    @hash.delete(object)
-    self
+  def delete(object) : T?
+    return nil if !object.is_a? T
+    @hash.delete(object) { return nil }
+    object
```
Comment (jhass) [inline src/set.cr:150]: Returning the argument as is indeed has no benefit to just returning `Bool`.

The potential advantage of returning `T?` is to return the *original object* that was in the set. This can be useful if you have objects with partial equality, that is you can reconstruct it based on data you have and that makes it equal (say based on opaque identifier  given to API consumers), while you need additional context data contained in the original object but not available in the current context (say a database PK).

Comment (j8r) [reply src/set.cr:150]: Yes that's right, I will change that

Comment (j8r) [reply src/set.cr]: That's not possible with the current `Hash` API, changing the method to return a `Bool`.

**Comment by j8r** [comment]
I have changed `Set#delete` to return `Bool`. It is not possible with the current `Hash` API to return the deleted key reference.

File: src/set.cr [hunk 6]
```diff
@@ -136,17 +136,18 @@ struct Set(T)
     @hash.has_key?(object)
   end
 
-  # Removes the *object* from the set and returns `self`.
+  # Removes the *object* and returns it, otherwise returns `nil`.
```
Comment (Sija) [inline src/set.cr:139]:
```suggestion
  # Removes the *object* from the set and returns `true`—if present, otherwise returns `false`.
```

**Comment by j8r** [comment]
Solves #9591

**Comment by j8r** [comment]
Closes #9591

File: src/set.cr [hunk 7]
```diff
@@ -136,7 +136,7 @@ struct Set(T)
     @hash.has_key?(object)
   end
 
-  # Removes the *object* and returns it, otherwise returns `nil`.
+  # Removes the *object* from the set and returns `true`—if present, otherwise returns `false`.
```
Comment (jhass) [inline src/set.cr:139]:
```suggestion
  # Removes the *object* from the set and returns `true` if it was present, otherwise returns `false`.
```

**Review by straight-shoota — APPROVED** [review]

File: spec/std/set_spec.cr [hunk 8]
```diff
@@ -63,9 +63,14 @@ describe "Set" do
       set.includes?(3).should be_true
     end
 
-    it "returns self" do
+    it "returns the object is present" do
```
Comment (jhass) [inline spec/std/set_spec.cr:66]:
```suggestion
    it "returns true when the object was present" do
```

File: spec/std/set_spec.cr [hunk 9]
```diff
@@ -63,9 +63,14 @@ describe "Set" do
...
+      set = Set{1, 2, 3}
+      set.delete(2).should be_true
+    end
+
+    it "returns nil if the object is absent" do
```
Comment (jhass) [inline spec/std/set_spec.cr:71]:
```suggestion
    it "returns false when the object was absent" do
```

**Review by jhass — APPROVED** [review]
Fixed outdated specs wording, looks good otherwise

**Review by straight-shoota — APPROVED** [review]
